File: algorithms/algorithms-cplusplus/leetcode/Question_0797_All_Paths_From_Source_to_Target.cc
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
    vector<vector<int>> paths;
    vector<int> path;
    dfs(graph, paths, path, 0);
    return paths;
  }

 private:
  void dfs(vector<vector<int>>& graph, vector<vector<int>>& res, vector<int>& path, int cur) {
    path.push_back(cur);
    if (cur == graph.size() - 1) {
      res.push_back(path);
    } else {
      for (int g : graph[cur]) {
        dfs(graph, res, path, g);
      }
    }
    path.pop_back();
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0797_All_Paths_From_Source_to_Target.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0797_All_Paths_From_Source_to_Target {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> paths = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(graph, 0, path, paths);
        return paths;
    }

    private void dfs(int[][] graph, int i, List<Integer> path, List<List<Integer>> paths) {
        path.add(i);
        if (i == graph.length - 1) {
            paths.add(new ArrayList<>(path));
        } else {
            for (int j : graph[i]) {
                dfs(graph, j, path, paths);
            }
        }
        path.remove(path.size() - 1);
    }
}
